### ISSUE TITLE
An example that runs on llvm backend but does not on the Haskell backend. 

### ISSUE THREAD

**Issue by ChristianoBraga** [kframework/kore#2995]
Guys, I know you are all overloaded with work, over 200 issues! But please take a look at the situation below.

The following K description runs just fine in the llvm backend but either loops or takes quite a while to finish in the Haskell backend.
This example is a simplification of Plutus semantics in K that arose while running the ML Proof Checker. Could you shed some light?
```
require "domains.md"
module TEST
 imports LIST
 imports K-EQUAL
 imports INT
 imports STRING

 syntax String ::= head(String) [function]
 rule head(S) => substrString(S, 0, 1)
 syntax String ::= tail(String) [function]
 rule tail(S) => substrString(S, 1, lengthString(S))

 syntax List ::= f(String, List) [function]
 rule f(S, L) =>
      #if S ==String ""
      #then (L)
      #else f(tail(S), (L ListItem(head(S))))
      #fi
endmodule
```

In LLVM:
```
$ kompile --backend llvm t.k --main-module TEST
[Warning] Compiler: Could not find main syntax module with name TEST-SYNTAX in
definition.  Use --syntax-module to specify one. Using TEST as default.
$ time krun -cPGM="f(\"ola\", .List)" --depth 0
<k>
  ListItem ( "o" )
  ListItem ( "l" )
  ListItem ( "a" ) ~> .
</k>

real    0m0.990s
user    0m2.462s
sys     0m0.133s
```

In Haskell:
```
$ kompile --backend haskell t.k --main-module TEST
[Warning] Compiler: Could not find main syntax module with name TEST-SYNTAX in
definition.  Use --syntax-module to specify one. Using TEST as default.

$ time krun -cPGM="f(\"ola\", .List)" --depth 0
^Ckore-exec: [12129483] Error (ErrorException):
    Error while communicating with the solver:
        fd:137: hFlush: resource vanished (Broken pipe)
    Solver exit code: -2
Created bug report: kore-exec.tar.gz
[Error] Critical: Backend crashed during rewriting with exit code 1

real    0m13.209s
user    0m16.102s
sys     0m0.606s
```

**Comment by dwightguth** [comment]
I believe the issue is that #if is not short circuiting, so it is always evaluating the recursive call even if the string is empty. This is not really a simple fix at all, even though it is unintuitive. I would recommend rewriting that rule as two rules with a side condition rather than a single rule with `#if`.

This is probably not going to get fixed anytime soon.

**Comment by ChristianoBraga** [comment]
Thank you

**Comment by ana-pantilie** [comment]
Not clear if there's a sound way to solve this.

@xc93 Do you have any idea on how to do definedness checking on this?

**Comment by virgil-serbanuta** [comment]
Would this work and be good enough (I may have misplaced some parenthesis)?
```
rule f(S, L) => (((S ==String "") #Equals true) #And (L)) #Or (#Not ((S ==String "") #Equals true) #And f(tail(S), (L ListItem(head(S)))))
```
This could be made to be more friendly by introducing a new ML operator/macro `#IF` which would be defined by `#IF(C, TB, FB) === (C #And TB) #Or ((#Not C) #And FB)` and which would be evaluated without expansion, in the expected way, by the Haskell backend.

**Comment by ehildenb** [comment]
Inductively/co-inductively discharge definedness check?

As I understand the problem, we have the rule:

```
rule f(S, L) => #if ... #then ... #else f(S', L') #fi
```

Then, it's trying to check `#Ceil(#if .... #then ... #else f(S', L') #fi)` in the case where the boolean condition is true (which it should have to do for soundness, even if evaluation is short-circuiting).

But in checking this `#Ceil(...)` conditino, it must also check `#Ceil(f(S', L'))`, which in needs to evaluate `f(S', L')` a bit to do so, which leads to applying the rule again, which leads to yet another definedness check of `f(S'', L'')`.

Is there a way we can coinductively or inductively discharge this definedness check? So once we go to check `#Ceil(f(S', L'))`, we say "this is an instance of what we're already trying to prove, which is `#Ceil(f(S, L))`, and so we don't need to continue unfolding the `#Ceil` or the `f` here". @xc93 is this sound? If so, can you write a proof of it? This would allow us to use short-circuiting operators in Haskell backend in a safe manner.

**Comment by dwightguth** [comment]
If you have an encoding that works for the haskell backend and provides the same results on all concrete inputs as the llvm backend, I would recommend simply writing a compilation pass in the k frontend that transforms `#if` into that encoding and running it on the haskell backend only.

**Comment by ehildenb** [comment]
@xc93 with this encoding, is it still true that `#if ...` is a `function` symbol or not?

If it's not the case, then maybe we cannot using the existing equational simplification engine of the Haskell backend for simplifying rules with `#if ...` at the top.

**Comment by dwightguth** [comment]
It is not, under this encoding, because it's not a symbol. It's a macro for an \or.

**Comment by xc93** [comment]
`#IF` is not a `function` symbol. In fact, it is not a symbol at all, but a notation. The reason is that even if the first argument is #BOTTOM, the result is not necessarily `#BOTTOM`. It is `FB`. 
`#BIF`, however, does have a `function` symbol semantics. Instead of defining `#BIF` as a notation using `#IF`, I can instead define it directly as a `function` symbol with two axioms as follows:
```
symbol BIF(Bool, K, K) : K [function]
axiom BIF(true, k1, k2) = k1
axiom BIF(false, k2, k2) = k2
```
The two axioms guarantee that for all `b, k1, k2`, `BIF(b, k1, k2)` is a singleton pattern, matched by precisely one element (either k1 or k2, depending on b). That is, `BIF` is a function from `Bool \times K \times K` to `K`.

**Comment by dwightguth** [comment]
For what it's worth, it was never treated as a regular function symbol by the llvm backend, any more than the other short circuiting boolean operators were. They were all campiled down to native branch instructions. So I don't see why we can't treat them specially in the haskell backend also.

**Comment by dwightguth** [comment]
@xc93 that doesn't sound correct; again, if k1 is bottom but b is false, the result is k2. This is not true if if is a symbol.

**Comment by xc93** [comment]
@dwightguth You are right. I take that back. Neither `#IF` nor `#BIF` can be symbols with their intended semantics.

**Comment by ehildenb** [comment]
A better K semantics wolud then be?

```
syntax {Sort} Sort ::= "#IF" Sort "#THEN" Sort "#ELSE" Sort "#FI" [hook(KEQUAL.mlif)] // Matching-logic #ite
rule #IF S #then B1 #else B2 #fi => (S #And B1) #Or ((#Not S) #And B2) [anywhere]

syntax {Sort} Sort ::= "#if" Bool "#then" Sort "#else" Sort "#fi" [hook(KEQUAL.bif), smt-hook(ite)]
rule #if C #then B1 #else B2 #fi => #IF {C #Equals true} #THEN B1 #ELSE B2 #FI [anywhere]
```

**Comment by ehildenb** [comment]
But this would only be valid encoding to Z3 I think if we first check definedness of the entire `#ite`

**Comment by xc93** [comment]
> So I see the definition here https://github.com/runtimeverification/k/blob/master/k-distribution/include/kframework/builtin/domains.md#equality-and-conditionals:
> 
> ```
>   syntax {Sort} Sort ::= "#if" Bool "#then" Sort "#else" Sort "#fi"   [function, functional, smt-hook(ite), hook(KEQUAL.ite)]
> ```
> 
> So this is incorrect ^^^? We need to remove `function` and `functional` attributes here. Also, it's not clear that `ite` encoding to Z3 is correct encoding anymore? Or, if the term is defined, then it's the correct encoding?
> 
> Also, what should the Haskell backend do then? If it sees `#ite` symbol, it should delay checking definedness until it evaluates the condition? I think that should happen if we remove `function` and `functional` attributes, but I'm not sure.

The [function, functional] attributes are incorrect. As @dwightguth said, we want `#ite(true, 1, #BOTTOM)` to be `1` and `#ite(false, #BOTTOM, 2)` to be `2`, meaning that `#ite` cannot be a symbol. It has to be a notation. 

Here's another question about using Bool as the condition sort though. What is the intended semantics of `#ite(#BOTTOM{Bool}, K1, K2)` and `#ite(#TOP{Bool}, K1, K2)`?

As for checking definedness of `#ite`, there are two things we can do. If the condition is concrete, we can evaluate it and reduce the problem into checking one of the branches (i.e., short-circuiting). Another way is to check the definedness of _both_ branches. If both are defined, then `#ite(c, K1, K2)` is defined, given that `c` will always be evaluated to `true` or `false` (and not #BOTTOM{Bool} or #TOP{Bool}).

**Comment by ehildenb** [comment]
Right, currently we do check the definedness of both branches, which is causing the original problem because that definedness check requires recursively evaluating a function on itself (that's what's manifesting as the original bug, or at least I thought). The current algorithm that the Haskell backend uses for doing this seems to end up looping. At least, that's my understanding of the issue, also based on other discussions we've had about short-circuiting operators.

So we need to figure out how to (i) do the short-circuiting, but also (ii) do the definedness check for this term that the Haskell backend uses to check state feasibility. It seems to me that (i) is solved by @xc93 identifying that it's logically sound to short-circuit simplification of this symbol.

I proposed here a way of discharging the definedness condition check in this looping case: https://github.com/runtimeverification/haskell-backend/issues/2995#issuecomment-1123924812. Maybe it's enough to just make the definedness check discharge more cleverly in this recursive case?

**Comment by dwightguth** [comment]
There is the issue of how to encode and handle the changes you are making in the backend, but I would caution you against making too many changes to the user-facing syntax declaration because that is likely to have a lot more user impact and also impact on the rest of the frontend and on the llvm backend. Like I said, I would instead recommend compiling the symbols in the frontend into whatever kore notation you want to use, probably late in the frontend compilation process.

**Comment by virgil-serbanuta** [comment]
> So we need to figure out how to (i) do the short-circuiting, but also (ii) do the definedness check for this term that the Haskell backend uses to check state feasibility. It seems to me that (i) is solved by @xc93 identifying that it's logically sound to short-circuit simplification of this symbol.

Is it logically sound to short-circuit a symbol? It seemed to me that @xc93 said that "Neither #IF nor #BIF can be symbols" and " #ite cannot be a symbol. It has to be a notation." What did I miss?

**Comment by ehildenb** [comment]
Sorry @virgil-serbanuta , I meant notation, not symbol.
